iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -180,35 +181,72 @@ public virtual bool TryGetValue<T>(string componentName, string propertyName, ou
                     $"TryGetValue will attempt to get the property value but may not behave as expected.", nameof(TryGetValue));
             }
 
+            // If either the component name or the property name is null, empty or whitespace,
+            // then return unsuccessfully (false) with the default value of the type <T> passed in.
+            if (string.IsNullOrWhiteSpace(componentName) || string.IsNullOrWhiteSpace(propertyName))
+            {
+                propertyValue = default;
+                return false;
+            }
+
             if (Contains(componentName, propertyName))
             {
                 object componentProperties = Collection[componentName];
 
+                // If the ClientPropertyCollection was constructed by the user application (eg. for updating the client properties)
+                // then the componentProperties are retrieved as a dictionary.
+                // The required property value can be fetched from the dictionary directly.
                 if (componentProperties is IDictionary<string, object> nestedDictionary)
                 {
-                    if (nestedDictionary.TryGetValue(propertyName, out object dictionaryElement))
+                    // First verify that the retrieved dictionary contains the component identifier { "__t": "c" }.
+                    // If not, then the retrieved nested dictionary is actually a root-level property of type map.
+                    if (nestedDictionary.TryGetValue(ConventionBasedConstants.ComponentIdentifierKey, out object componentIdentifierValue)
+                        && componentIdentifierValue.Equals(ConventionBasedConstants.ComponentIdentifierValue))
                     {
-                        // If the value is null, go ahead and return it.
-                        if (dictionaryElement == null)
-                        {
-                            propertyValue = default;
-                            return true;
-                        }
-
-                        // If the object is of type T or can be cast to type T, go ahead and return it.
-                        if (dictionaryElement is T valueRef
-                            || NumericHelpers.TryCastNumericTo(dictionaryElement, out valueRef))
+                        if (nestedDictionary.TryGetValue(propertyName, out object dictionaryElement))
                         {
-                            propertyValue = valueRef;
-                            return true;
+                            // If the value associated with the key is null, then return successfully (true) with the default value of the type <T> passed in.
```

Comment:
I'm not sure if you need to say successfully. Just saying true/false makes it less confusing as the return value is a bool.